main.py

```diff
@@ -0,0 +1,277 @@
+        momentum=config["momentum"],
+        weight_decay=config["weight_decay"],
+    )
+    lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(
+        optimizer, milestones=config["lr_steps"], gamma=config["lr_gamma"]
+    )
+    start_epoch = 0
+
+    checkpoint = session.get_checkpoint()
+    if checkpoint is not None:
+        checkpoint_data = checkpoint.to_dict()
+        model.load_state_dict(checkpoint_data["model"])
+        optimizer.load_state_dict(checkpoint_data["optimizer"])
+        lr_scheduler.load_state_dict(checkpoint_data["lr_scheduler"])
+        start_epoch = checkpoint_data["epoch"]
+
+    for epoch in range(start_epoch, config["epochs"]):
+        train_one_epoch(
+            model=model,
+            optimizer=optimizer,
+            batch_size=config["batch_size"],
+            epoch=epoch,
+        )
+        lr_scheduler.step()
+        checkpoint = Checkpoint.from_dict(
+            {
+                "model": model.module.state_dict(),
+                "optimizer": optimizer.state_dict(),
+                "lr_scheduler": lr_scheduler.state_dict(),
+                "config": config,
+                "epoch": epoch,
+            }
+        )
+        session.report({}, checkpoint=checkpoint)
+
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_loop_per_worker,
+    train_loop_config={
+        "batch_size": 2,
+        "lr": 0.02,
+        # "epochs": 26,
+        "epochs": 1,
+        "momentum": 0.9,
+        "weight_decay": 1e-4,
+        "lr_steps": [16, 22],
+        "lr_gamma": 0.1,
+    },
+    scaling_config=ScalingConfig(num_workers=2, use_gpu=False),
+    # run_config=RunConfig(progress_reporter=CLIReporter)  TODO
+    datasets={"train": train_dataset.limit(100)},
+)
+results = trainer.fit()
+
+
+class CustomTorchPredictor(TorchPredictor):
+    def _predict_numpy(
+        self, data: np.ndarray, dtype: torch.dtype
+    ) -> Dict[str, np.ndarray]:
+        inputs = [torch.as_tensor(image) for image in data["image"]]
+        assert all(image.dim() == 3 for image in inputs)
+        outputs = self.call_model(inputs)
+
+        predictions = collections.defaultdict(list)
+        for output in outputs:
+            for key, value in output.items():
+                predictions[key].append(value.cpu().detach().numpy())
+
+        for key, value in predictions.items():
+            predictions[key] = _create_possibly_ragged_ndarray(value)
+        predictions = {"pred_" + key: value for key, value in predictions.items()}
+        return predictions
+
+
+model = models.detection.fasterrcnn_resnet50_fpn(num_classes=21)
+
+predictor = BatchPredictor(
+    results.checkpoint, CustomTorchPredictor, model=model, use_gpu=False
+)
+predictions = predictor.predict(
+    test_dataset.limit(100),
+    feature_columns=["image"],
+    keep_columns=["boxes", "labels"],
+    batch_size=4,
+    max_scoring_workers=1
+)
+
+
+metric = MeanAveragePrecision()
+for row in predictions.iter_rows():
```

Comment:
can we use map_batches here instead of iter_rows? iter_rows won't really scale if the data is too large.

Would we be able to use `AveragePrecision` in the `map_batches` and then do the mean ourselves at the end?

Reply:
Yeah, I'll add a more scalable implementation. 

IIRC we can't use `AveragePrecision` because it's intended for classification and doesn't work with object detection data.